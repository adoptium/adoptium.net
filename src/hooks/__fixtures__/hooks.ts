import {
  LatestTemurin,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  ReleaseNoteAPIResponse,
  ReleaseAsset,
  TemurinReleases,
} from "@/hooks";

export const createRandomTemurinRelease = (
  installer: boolean
): ReleaseAsset => ({
  os: "os_mock",
  architecture: "arch_mock",
  release_link: "https://release_link_mock",
  platform_name: "platform_name_mock",
  release_name: "release_name_mock",
  release_date: "2020-01-01T00:00:00Z",
  version: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.1+0",
  },
  binaries: [
    {
      type: "type_mock",
      link: "https://link_mock",
      checksum: "checksum_mock",
      size: 0,
      extension: "extension_mock",
      release_name: "release_name_mock",
      installer_link: installer ? "https://installer_link_mock" : undefined,
      installer_checksum: installer ? "installer_checksum_mock" : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? "installer_extension_mock" : undefined,
    },
  ],
});

export const mockLatestTemurin = (installer: boolean): LatestTemurin => ({
  download_count: 0,
  id: "id_mock",
  release_link: "https://release_link_mock",
  release_name: "release_name_mock",
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.1+0",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "type_mock",
      heap_size: "heap_size_mock",
      download_count: 0,
      jvm_impl: "jvm_impl_mock",
      package: {
        name: "name_mock.tar.gz",
        link: "https://link_mock",
        checksum: "sha256sum_mock",
        checksum_link: "https://sha256sum_link_mock",
        metadata_link: "https://metadata_link_mock",
        signature_link: "https://signature_link_mock",
        size: 0,
        download_count: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            link: "https://installer_link_mock",
            checksum: "installer_sha256sum_mock",
            checksum_link: "https://installer_sha256sum_link_mock",
            metadata_link: "https://installer_metadata_link_mock",
            signature_link: "https://installer_signature_link_mock",
            size: 0,
            download_count: 0,
          }
        : undefined,
    },
  ],
});

export const createMockTemurinReleaseAPI = (
  installer: boolean,
  image_type: string
): MockTemurinReleaseAPI => ({
  release_link: "https://release_link_mock",
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  version: {
    major: 21,
    minor: 0,
    security: 12,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "21.0.12+0",
  },
  binary: {
    architecture: "arch_mock",
    project: "project_mock",
    scm_ref: "scm_ref_mock",
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: "os_mock",
    image_type: image_type ? image_type : "jdk",
    jvm_impl: "jvm_impl_mock",
    heap_size: "heap_size_mock",
    package: {
      name: "name_mock.tar.gz",
      link: "https://link_mock",
      checksum: "checksum_mock",
      checksum_link: "https://checksum_link_mock",
      metadata_link: "https://metadata_link_mock",
      signature_link: "https://signature_link_mock",
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: "installer_name_mock.msi",
          link: "https://installer_link_mock",
          checksum: "installer_checksum_mock",
          checksum_link: "https://installer_checksum_link_mock",
          metadata_link: "https://installer_metadata_link_mock",
          signature_link: "https://installer_signature_link_mock",
          size: 0,
          download_count: 0,
        }
      : undefined,
  },
});

export const createMockReleaseNotesAPI = (
  number: number
): ReleaseNoteAPIResponse => ({
  id: "id_mock",
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "0.0.0+0",
  },
  release_name: "release_name_mock",
  // return an array of length number with the same release notes
  release_notes: Array.from({ length: number }, (v, i) => ({
    id: i.toString(),
    link: new URL("https://link_mock"),
    title: "title_mock",
    priority: "1",
    type: "Bug",
    component: "component_mock",
  })),
});

export const createMockTemurinFeatureReleaseAPI = (
  installer: boolean
): MockTemurinFeatureReleaseAPI => ({
  id: "id_mock",
  download_count: 0,
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
    semver: "0.0.0+0",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "jdk",
      package: {
        name: "name_mock.tar.gz",
        checksum: "checksum_mock",
        link: new URL("https://link_mock"),
        size: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            checksum: "installer_checksum_mock",
            link: new URL("https://installer_link_mock"),
            size: 0,
          }
        : undefined,
    },
  ],
});

export const createRandomTemurinReleases = (
  installer: boolean
): TemurinReleases => ({
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  source_url: new URL("https://source_url_mock"),
  release_notes: true,
  release_notes_name: "release_notes_name_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    platform_mock: {
      assets: [
        {
          os: "os_mock",
          architecture: "architecture_mock",
          type: "type_mock",
          link: new URL("https://link_mock"),
          checksum: "checksum_mock",
          size: 0,
          extension: "extension_mock",
          installer_link: installer
            ? new URL("https://installer_link_mock")
            : undefined,
          installer_checksum: installer ? "installer_checksum_mock" : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer
            ? "installer_extension_mock"
            : undefined,
        },
      ],
    },
  },
});
