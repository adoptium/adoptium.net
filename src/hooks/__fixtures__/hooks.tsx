import {
  Binary,
  Contributor,
  ContributorApiResponse,
  EventAPI,
  LatestTemurin,
  MarketplaceRelease,
  MockTemurinFeatureReleaseAPI,
  MockTemurinReleaseAPI,
  ReleaseNoteAPIResponse,
  News,
  NewsResponse,
  ReleaseAsset,
  TemurinReleases,
  OperatingSystem,
  Architecture
} from "../hooks"

export const createRandomTemurinRelease = (installer): ReleaseAsset => ({
  os: "os_mock",
  architecture: "arch_mock",
  release_link: new URL("https://release_link_mock"),
  platform_name: "platform_name_mock",
  release_name: "release_name_mock",
  release_date: new Date(Date.UTC(2020, 0, 1)),
  binaries: [
    {
      type: "type_mock",
      link: new URL("https://link_mock"),
      checksum: "checksum_mock",
      size: 0,
      extension: "extension_mock",
      installer_link: installer
        ? new URL("https://installer_link_mock")
        : undefined,
      installer_checksum: installer ? "installer_checksum_mock" : undefined,
      installer_size: installer ? 0 : undefined,
      installer_extension: installer ? "installer_extension_mock" : undefined,
    },
  ],
})

export const createRandomMarketplaceRelease = (
  installer,
  id,
): MarketplaceRelease => ({
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  binary: {
    os: "os_mock",
    architecture: "arch_mock",
    distribution: "distribution_mock",
    image_type: "type_mock",
    jvm_impl: "jvm_impl_mock",
    timestamp: new Date(Date.UTC(2020, 0, 1)),
    scm_ref: "scm_ref_mock",
    openjdk_scm_ref: "openjdk_scm_ref_mock",
    aqavit_results_link: new URL("https://aqavit_results_link_mock"),
    tck_affidavit_link: new URL("https://tck_affidavit_link_mock"),
    package: {
      name: "name_mock.tar.gz",
      link: new URL("https://link_mock"),
      sha256sum: `sha256sum_mock${id}`,
      sha256sum_link: new URL("https://sha256sum_link_mock"),
      signature_link: new URL("https://signature_link_mock"),
    },
    installer: installer
      ? [
          {
            name: "installer_name_mock.msi",
            link: new URL("https://installer_link_mock"),
            sha256sum: "installer_sha256sum_mock",
            sha256sum_link: new URL("https://installer_sha256sum_link_mock"),
            signature_link: new URL("https://installer_signature_link_mock"),
          },
        ]
      : undefined,
  },
  version: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
})

export const mockLatestTemurin = (installer): LatestTemurin => ({
  download_count: 0,
  id: "id_mock",
  release_link: new URL("https://release_link_mock"),
  release_name: "release_name_mock",
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 21,
    minor: 0,
    security: 1,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "type_mock",
      heap_size: "heap_size_mock",
      download_count: 0,
      jvm_impl: "jvm_impl_mock",
      package: {
        name: "name_mock.tar.gz",
        link: new URL("https://link_mock"),
        checksum: "sha256sum_mock",
        checksum_link: new URL("https://sha256sum_link_mock"),
        metadata_link: new URL("https://metadata_link_mock"),
        signature_link: new URL("https://signature_link_mock"),
        size: 0,
        download_count: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            link: new URL("https://installer_link_mock"),
            checksum: "installer_sha256sum_mock",
            checksum_link: new URL("https://installer_sha256sum_link_mock"),
            metadata_link: new URL("https://installer_metadata_link_mock"),
            signature_link: new URL("https://installer_signature_link_mock"),
            size: 0,
            download_count: 0,
          }
        : undefined,
    },
  ],
})

export const mockNewsAPI = (): NewsResponse => ({
  news: [
    {
      title: "news_title_mock",
      body: "body_mock",
      id: "id_mock",
      link: new URL("https://link_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
  pagination: {
    page: 1,
    pagesize: 5,
    result_start: 1,
    result_end: 5,
    result_size: 5,
    total_result_size: 16,
  },
})

export const mockEventsAPI = (): EventAPI => ({
  events: [
    {
      title: "events_title_mock",
      id: "id_mock",
      infoLink: new URL("https://link_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
})

export const createMockTemurinReleaseAPI = (
  installer,
  image_type,
): MockTemurinReleaseAPI => ({
  release_link: new URL("https://release_link_mock"),
  release_name: "release_name_mock",
  vendor: "vendor_mock",
  version: {
    major: 21,
    minor: 0,
    security: 12,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binary: {
    architecture: "arch_mock",
    project: "project_mock",
    scm_ref: "scm_ref_mock",
    download_count: 0,
    updated_at: new Date(Date.UTC(2020, 0, 1)),
    os: "os_mock",
    image_type: image_type ? image_type : "jdk",
    jvm_impl: "jvm_impl_mock",
    heap_size: "heap_size_mock",
    package: {
      name: "name_mock.tar.gz",
      link: new URL("https://link_mock"),
      checksum: "checksum_mock",
      checksum_link: new URL("https://checksum_link_mock"),
      metadata_link: new URL("https://metadata_link_mock"),
      signature_link: new URL("https://signature_link_mock"),
      size: 0,
      download_count: 0,
    },
    installer: installer
      ? {
          name: "installer_name_mock.msi",
          link: new URL("https://installer_link_mock"),
          checksum: "installer_checksum_mock",
          checksum_link: new URL("https://installer_checksum_link_mock"),
          metadata_link: new URL("https://installer_metadata_link_mock"),
          signature_link: new URL("https://installer_signature_link_mock"),
          size: 0,
          download_count: 0,
        }
      : undefined,
  },
})

export const createMockReleaseNotesAPI = (number): ReleaseNoteAPIResponse => ({
  id: "id_mock",
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  release_name: "release_name_mock",
  // return an array of length number with the same release notes
  release_notes: Array.from({ length: number }, (v, i) => ({
    id: i.toString(),
    link: new URL("https://link_mock"),
    title: "title_mock",
    priority: "1",
    type: "Bug",
    component: "component_mock",
  })),
})

export const createMockTemurinFeatureReleaseAPI = (
  installer,
): MockTemurinFeatureReleaseAPI => ({
  id: "id_mock",
  download_count: 0,
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  release_type: "release_type_mock",
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  updated_at: new Date(Date.UTC(2020, 0, 1)),
  vendor: "vendor_mock",
  version_data: {
    major: 0,
    minor: 0,
    security: 0,
    patch: 0,
    build: 0,
    openjdk_version: "openjdk_version_mock",
  },
  binaries: [
    {
      os: "os_mock",
      architecture: "arch_mock",
      image_type: "jdk",
      package: {
        name: "name_mock.tar.gz",
        checksum: "checksum_mock",
        link: new URL("https://link_mock"),
        size: 0,
      },
      installer: installer
        ? {
            name: "installer_name_mock.msi",
            checksum: "installer_checksum_mock",
            link: new URL("https://installer_link_mock"),
            size: 0,
          }
        : undefined,
    },
  ],
})

export const createRandomTemurinReleases = (installer): TemurinReleases => ({
  release_name: "release_name_mock",
  release_link: new URL("https://release_link_mock"),
  source_url: new URL("https://source_url_mock"),
  release_notes: true,
  release_notes_name: 'release_notes_name_mock',
  timestamp: new Date(Date.UTC(2020, 0, 1)),
  platforms: {
    platform_mock: {
      assets: [
        {
          os: "os_mock",
          architecture: "architecture_mock",
          type: "type_mock",
          link: new URL("https://link_mock"),
          checksum: "checksum_mock",
          size: 0,
          extension: "extension_mock",
          installer_link: installer
            ? new URL("https://installer_link_mock")
            : undefined,
          installer_checksum: installer ? "installer_checksum_mock" : undefined,
          installer_size: installer ? 0 : undefined,
          installer_extension: installer
            ? "installer_extension_mock"
            : undefined,
        },
      ],
    },
  },
})

export const createRandomContributorViewData = (): Contributor => ({
  avatarUri: "avatarUri_mock",
  profileUri: "profileUri_mock",
  login: "login_mock",
  contributionsCount: 0,
  commitsListUri: "commitsListUri_mock",
  repo: "reponame_mock",
})

export const createRandomLatestForOSData = (): Binary => ({
  release_name: "release_name_mock",
  link: new URL("https://link_mock"),
})

export const createRandomNewsAndEventsData = (): News => ({
  news: {
    news: [
      {
        id: "id_mock",
        title: "title_mock",
        body: "body_mock",
        date: new Date(Date.UTC(2020, 0, 1)),
        link: new URL("https://link_mock"),
      },
    ],
    pager: {     
      current_page: 0,
      items_per_page: 5,
      total_items: 54,
      total_pages: 11,
    }
  },
  events: [
    {
      id: "id_mock",
      title: "title_mock",
      infoLink: new URL("https://infoLink_mock"),
      date: new Date(Date.UTC(2020, 0, 1)),
    },
  ],
})

export const createRandomContributorViewData1Contribution =
  (): Contributor => ({
    avatarUri: "avatarUri_mock",
    profileUri: "profileUri_mock",
    login: "login_mock",
    contributionsCount: 1,
    commitsListUri: "commitsListUri_mock",
    repo: "reponame_mock",
  })

export const createMockAdoptiumContributorsApi =
  (): ContributorApiResponse => ({
    login: "test-user",
    id: 1,
    url: "https://api.github.com/users/test-user",
    type: "User",
    contributions: 10,
    node_id: "1",
    avatar_url: "https://github.com/images/error/octocat_happy.gif",
    gravatar_id: "",
    html_url: "https://github.com/images/error/octocat_happy.gif",
    followers_url: "https://api.github.com/users/test-user/followers",
    following_url:
      "https://api.github.com/users/test-user/following{/other_user}",
    gists_url: "https://api.github.com/users/test-user/gists{/gist_id}",
    starred_url:
      "https://api.github.com/users/test-user/starred{/owner}{/repo}",
    subscriptions_url: "https://api.github.com/users/test-user/subscriptions",
    organizations_url: "https://api.github.com/users/test-user/orgs",
    repos_url: "https://api.github.com/users/test-user/repos",
    events_url: "https://api.github.com/users/test-user/events{/privacy}",
    received_events_url:
      "https://api.github.com/users/test-user/received_events",
    site_admin: false,
  })

  export const mockOsesAPI = (): OperatingSystem[] => (
    [{name: "mock_macos", value: "mock_macos"}, {name: "mock_linux", value: "mock_linux"}, {name: "mock_windows", value: "mock_windows"}]
  )
  
  export const mockArchesAPI = (): Architecture[] => (
    [{name: "mock_aarch64", value: "mock_aarch64"}, {name: "mock_ppc64", value: "mock_ppc64"}, {name: "mock_x64", value: "mock_x64"}
  ])
  