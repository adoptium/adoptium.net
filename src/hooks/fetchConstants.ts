import { useEffect, useState } from "react";
import axios from "axios";
import { capitalize } from "@/utils/capitalize";

const baseUrl = "https://api.adoptium.net/v3";

export function useOses(isVisible: boolean): OperatingSystem[] {
  const [oses, setOses] = useState<OperatingSystem[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/operating_systems`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newOses = response.data
              .map((s: string) => {
                const o: OperatingSystem = {
                  name: capitalize(s),
                  value: s.toLowerCase(),
                };
                return o;
              })
              .sort((o1: OperatingSystem, o2: OperatingSystem) =>
                o1.name.localeCompare(o2.name)
              );

            setOses(newOses);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setOses([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return oses;
}

export function useArches(isVisible: boolean): Architecture[] {
  const [arches, setArches] = useState<Architecture[]>([]);

  useEffect(() => {
    let isMounted = true;
    const cancelToken = axios.CancelToken.source();

    if (isVisible) {
      (async () => {
        const url = `${baseUrl}/types/architectures`;

        axios
          .get(url, { cancelToken: cancelToken.token })
          .then(function (response) {
            if (!isMounted) return;

            const newArches = response.data
              .map((s: string) => {
                const a: Architecture = { name: s, value: s.toLowerCase() };
                if (a.name === "x32") a.name = "x86";
                return a;
              })
              .sort((a1: Architecture, a2: Architecture) =>
                a1.name.localeCompare(a2.name)
              );

            setArches(newArches);
          })
          .catch(function (error) {
            if (!isMounted || axios.isCancel(error)) return;
            setArches([]);
          });
      })();
    }

    return () => {
      isMounted = false;
      cancelToken.cancel("Component unmounted");
    };
  }, [isVisible]);

  return arches;
}

export interface OperatingSystem {
  name: string;
  value: string;
}

export interface Architecture {
  name: string;
  value: string;
}
