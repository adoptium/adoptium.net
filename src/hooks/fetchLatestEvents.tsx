import axios from "axios"

const baseUrl = "https://newsroom.eclipse.org/api"

export async function fetchLatestEvents() {
    const url = new URL(`${baseUrl}/events`)
    url.searchParams.append("parameters[publish_to]", "adoptium")
    url.searchParams.append("pagesize", "20")

    return axios
        .get(url.toString())
        .then(function (response) {
            return response.data.events
        })
        .catch(function () {
            return []
        })
}

export interface News {
    news: NewsResponse
    events: EventItem[]
}

export interface NewsResponse {
    news: NewsItem[]
    pager: {
        current_page: number
        items_per_page: number
        total_items: number
        total_pages: number
    } | null
}

export interface NewsItem {
    id: string
    title: string
    body: string
    date: Date
    link: URL
    image?: string  // Optional image URL from the Eclipse Foundation news API
}

export interface EventItem {
    id: string
    title: string
    infoLink: URL
    date: Date
}

export interface EventAPI {
    events: EventItem[]
}
