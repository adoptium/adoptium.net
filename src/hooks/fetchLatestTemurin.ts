'use client'

import { Binary, VersionMetaData } from "@/types/temurin"
import { useEffect, useState } from "react"
import axios from "axios"

const baseUrl = "https://api.adoptium.net/v3"

export interface LatestTemurin {
  download_count: number
  id: string
  release_link: string
  release_name: string
  release_type: string
  source?: {
    link: string
    name: string
    size: number
  }
  timestamp: Date
  updated_at: Date
  vendor: string
  version_data: VersionMetaData
  binaries: [
    {
      architecture: string
      download_count: number
      heap_size: string
      image_type: string
      jvm_impl: string
      os: string
      package: {
        checksum: string
        checksum_link: string
        download_count: number
        link: string
        metadata_link: string
        name: string
        signature_link: string
        size: number
      }
      installer?: {
        checksum: string
        checksum_link: string
        download_count: number
        link: string
        metadata_link: string
        name: string
        signature_link: string
        size: number
      }
    }
  ]
}

export function useFetchLatestForOS(
  isVisible: boolean,
  version: number | string,
  os: string,
  arch: string,
): Binary | null {
  // Always declare hooks at the top level
  const [binary, setBinary] = useState<Binary | null>(null)

  useEffect(() => {
    // Early return inside useEffect instead of before hooks
    if (!os || !isVisible) {
      return;
    }

    const fetchBinary = async () => {
      const url = `${baseUrl}/assets/feature_releases/${version}/ga?os=${os}&architecture=${arch}&image_type=jdk&jvm_impl=hotspot&page_size=1&vendor=eclipse`

      try {
        const response = await axios.get(url);
        const json: LatestTemurin = response.data[0];

        let binary_link = json.binaries[0].package.link;
        let binary_checksum = json.binaries[0].package.checksum;

        if (json.binaries[0].installer) {
          binary_link = json.binaries[0].installer.link;
          binary_checksum = json.binaries[0].installer.checksum;
        }

        const packageData = json.binaries[0].package;
        const binary: Binary = {
          release_name: json.release_name,
          link: binary_link,
          checksum: binary_checksum,
          type: json.binaries[0].image_type,
          size: packageData.size,
          extension: packageData.name.split('.').pop() || ''
        };
        setBinary(binary);
      } catch {
        // Just set binary to null on error
        setBinary(null);
      }
    };

    fetchBinary();
  }, [isVisible, version, os, arch]);

  return binary;
}
