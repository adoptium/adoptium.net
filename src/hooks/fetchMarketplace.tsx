import { VersionMetaData } from "."
import axios from "axios"
import { ReleaseAsset } from "@/types/temurin"

// Define a type that extends ReleaseAsset[] to include the binary property
export interface MarketplaceReleaseAsset extends ReleaseAsset {
    binary: {
        distribution: string
    }
}

const baseUrl = "https://marketplace-api.adoptium.net"

export async function getAllPkgsForVersion(
    version: number,
    os: string,
    architecture: string,
    package_type: string,
    vendors: string[] = [],
): Promise<MarketplaceRelease[] | null> {
    let params = "?"
    params += "feature_version=" + version

    if (os !== "any") {
        if (os === "alpine-linux") {
            params += "&os=alpine_linux"
        } else {
            params += "&os=" + os
        }
    }

    // loop through vendors and add to params
    if (vendors.length > 0) {
        vendors.forEach(vendor => {
            params += "&vendor=" + vendor
        })
    }

    if (package_type !== "any") {
        params += "&image_type=" + package_type
    }

    if (architecture !== "any") {
        params += "&architecture=" + architecture
    }

    const url = baseUrl + "/v1/assets/latestForVendors" + params
    const data = await axios
        .get(url)
        .then(function (response) {
            return response.data
        })
        .catch(function () {
            return null
        })

    return data
}

export function getImageForDistribution(distribution: string) {
    switch (distribution) {
        case "temurin":
            return "/images/vendors/adoptium-logo.png"
        case "redhat":
            return "/images/members/redhat.svg"
        case "zulu":
            return "/images/vendors/azul-logo.png"
        case "semeru":
            return "/images/members/ibm-logo.png"
        case "microsoft":
            return "/images/members/microsoft.svg"
        default:
            return `/images/vendors/${distribution}-logo.png`
    }
}

export interface MarketplaceRelease {
    release_name: string
    vendor: string
    binary: {
        os: string
        architecture: string
        image_type: string
        jvm_impl: string
        package: {
            name: string
            link: URL
            sha256sum: string
            sha256sum_link: URL
            signature_link: URL
        }
        installer?: [
            {
                name: string
                link: URL
                sha256sum: string
                sha256sum_link: URL
                signature_link: URL
            },
        ]
        timestamp: Date
        scm_ref: string
        openjdk_scm_ref: string
        distribution: string
        aqavit_results_link: URL
        tck_affidavit_link: URL
    }
    version: VersionMetaData
}
