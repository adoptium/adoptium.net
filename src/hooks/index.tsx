'use client'

export * from "./useOnScreen"
export * from "./fetchLatestTemurin"
export * from "./fetchTemurinReleases"
export * from "./fetchMarketplace"

export interface VersionMetaData {
    major: number
    minor: number
    security: number
    pre?: string
    patch?: number
    build: number
    optional?: string
    adopt_build_number?: number
    openjdk_version: string
    semver: string
}

/**
 * This method compares two versions and returns:
 * ->  0 if versions are equals
 * -> -1 if versionA is lower than versionB
 * ->  1 if versionA is highter than versionB
 *
 * @param versionA
 * @param versionB
 * @returns
 */
export function compareVersions(
    versionA: VersionMetaData,
    versionB: VersionMetaData,
): number {
    let comparison = versionA.major - versionB.major
    if (comparison === 0) {
        comparison = versionA.minor - versionB.minor
        if (comparison === 0) {
            comparison = versionA.security - versionB.security
            if (comparison === 0) {
                comparison = versionA.build - versionB.build
                if (comparison === 0) {
                    comparison =
                        (versionA.adopt_build_number ? versionA.adopt_build_number : 0) -
                        (versionB.adopt_build_number ? versionB.adopt_build_number : 0)
                }
            }
        }
    }
    return comparison
}

export function getVersionAsString(
    version: VersionMetaData,
    allowShortNotation: boolean = false,
): string {
    let result = `${version.major}.${version.minor}.${version.security}`
    if (
        allowShortNotation &&
        !version.minor &&
        !version.security &&
        !version.patch
    ) {
        // when minor, security & patch are 0 => do not print .0.0.0
        result = `${version.major}`
    }
    if (version.patch) {
        result += `.${version.patch}`
    }
    if (version.pre) {
        result += `-${version.pre}`
    }
    if (version.build) {
        result += `+${version.build}`
        if (version.adopt_build_number) result += `.${version.adopt_build_number}`
    }
    if (version.optional) {
        result += `-${version.optional}`
    }
    return result
}
