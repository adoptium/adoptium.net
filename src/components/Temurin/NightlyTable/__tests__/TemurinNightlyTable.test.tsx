import React from "react"
import { render } from "@testing-library/react"
import { describe, expect, it } from "vitest"
import { createRandomTemurinReleases } from "./__fixtures__/hooks"
import TemurinNightlyTable from ".."

const releases = {
  pagecount: null,
  releases: [
    createRandomTemurinReleases(false),
    createRandomTemurinReleases(true),
  ],
}

describe("TemurinNightlyTable component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <TemurinNightlyTable
        // @ts-expect-error - results is incomplete
        results={releases}
        openModalWithChecksum={() => { }}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it("renders correctly - no data", () => {
    const { container } = render(<TemurinNightlyTable
      // @ts-expect-error - results is invalid
      results={undefined}
      openModalWithChecksum={() => { }}
    />)
    expect(container).toMatchSnapshot()
  })
})
