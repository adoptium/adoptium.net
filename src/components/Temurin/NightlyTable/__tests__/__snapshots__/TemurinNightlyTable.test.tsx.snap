// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TemurinNightlyTable component > renders correctly - no data 1`] = `
<div>
  <div
    class="text-center text-gray-400 py-8"
  >
    Text
  </div>
</div>
`;

exports[`TemurinNightlyTable component > renders correctly 1`] = `
<div>
  <div
    class="overflow-x-auto rounded-xl bg-[#1B1132]"
  >
    <table
      class="min-w-full divide-y divide-[#3E3355]"
    >
      <thead
        class="bg-[#26193F]"
      >
        <tr>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            Text
          </th>
          <th
            class="px-6 py-3 text-left text-xs font-medium text-gray-400 uppercase tracking-wider"
          >
            SHA256
          </th>
        </tr>
      </thead>
      <tbody
        class="bg-[#1B1132] divide-y divide-[#3E3355]"
      >
        <tr>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Os_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            architecture_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Type_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-200"
          >
            release_name_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Jan 1, 2020
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-pink"
          >
            <a
              href="https://link_mock/"
              rel="noopener noreferrer"
              target="_blank"
            >
              extension_mock
            </a>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-pink"
          >
            <span
              class="text-gray-500"
            >
              Text
            </span>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            <button
              class="text-pink"
            >
              Text
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Os_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            architecture_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Type_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-200"
          >
            release_name_mock
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            Jan 1, 2020
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-pink"
          >
            <a
              href="https://link_mock/"
              rel="noopener noreferrer"
              target="_blank"
            >
              extension_mock
            </a>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-pink"
          >
            <a
              href="https://installer_link_mock/"
              rel="noopener noreferrer"
              target="_blank"
            >
              installer_extension_mock
            </a>
          </td>
          <td
            class="px-6 py-4 whitespace-nowrap text-sm text-gray-300"
          >
            <button
              class="text-pink"
            >
              Text
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
